core/src/main/java/net/kyori/moonshine/Moonshine.java

```diff
@@ -0,0 +1,205 @@
+   */
+  private final IPlaceholderResolverStrategy<R, I, F> placeholderResolverStrategy;
+
+  /**
+   * The source of intermediate messages, per receiver.
+   */
+  private final IMessageSource<R, I> messageSource;
+
+  /**
+   * The renderer of all messages, before sent via {@link #messageSender()}.
+   */
+  private final IMessageRenderer<R, I, O, F> messageRenderer;
+
+  /**
+   * The message sender of intermediate messages to a given receiver with resolved placeholders.
+   */
+  private final IMessageSender<R, O> messageSender;
+
+  /**
+   * A navigable set for iterating through the {@link IReceiverLocatorResolver}s with weight-based ordering.
+   */
+  private final NavigableSet<Weighted<? extends IReceiverLocatorResolver<? extends R>>> weightedReceiverLocatorResolvers;
+
+  /**
+   * A map of types to navigable sets for iterating through the {@link IPlaceholderResolver}s with weight-based
+   * ordering.
+   */
+  private final Map<Type, NavigableSet<Weighted<? extends IPlaceholderResolver<? extends R, ?, ? extends F>>>> weightedPlaceholderResolvers;
+
+  /**
+   * All scanned methods of this proxy, excluding special-case methods such as {@code default} methods and any returning
+   * {@link Moonshine}.
+   */
+  private final Map<Method, MoonshineMethod<? extends R>> scannedMethods;
+
+  Moonshine(final TypeToken<?> proxiedType,
+      final IPlaceholderResolverStrategy<R, I, F> placeholderResolverStrategy,
+      final IMessageSource<R, I> messageSource,
+      final IMessageRenderer<R, I, O, F> messageRenderer,
+      final IMessageSender<R, O> messageSender,
+      final NavigableSet<Weighted<? extends IReceiverLocatorResolver<? extends R>>> weightedReceiverLocatorResolvers,
+      final Map<Type, NavigableSet<Weighted<? extends IPlaceholderResolver<? extends R, ?, ? extends F>>>> weightedPlaceholderResolvers)
+      throws UnscannableMethodException {
+    this.proxiedType = proxiedType;
+    this.placeholderResolverStrategy = placeholderResolverStrategy;
+    this.messageSource = messageSource;
+    this.messageRenderer = messageRenderer;
+    this.messageSender = messageSender;
+    this.weightedReceiverLocatorResolvers = Collections.unmodifiableNavigableSet(weightedReceiverLocatorResolvers);
+    this.weightedPlaceholderResolvers = Collections.unmodifiableMap(weightedPlaceholderResolvers);
+
+    final Method[] methods = GenericTypeReflector.erase(proxiedType.getType()).getMethods();
+    final Map<Method, MoonshineMethod<? extends R>> scannedMethods = new HashMap<>(methods.length);
+    for (final Method method : methods) {
+      if (method.isDefault() || method.getReturnType() == Moonshine.class) {
+        continue;
+      }
+
+      final MoonshineMethod<? extends R> moonshineMethod =
+          new MoonshineMethod<>(this, proxiedType, method);
+      scannedMethods.put(method, moonshineMethod);
+    }
+    this.scannedMethods = Collections.unmodifiableMap(scannedMethods);
```

Comment:
Refactor this into a stream

Reply:
Not possible; `MoonshineMethod#<init>` can throw a checked exception.